internal/dataplane/parser/translators/grpcroute.go

```diff
@@ -48,6 +48,29 @@ func GenerateKongRoutesFromGRPCRouteRule(
 	// gather the k8s object information and hostnames from the grpcroute
 	ingressObjectInfo := util.FromK8sObject(grpcroute)
 
+	// generate a route to match hostnames only if there is no match in the rule.
+	if len(rule.Matches) == 0 {
+		// REVIEW: return an error here to tell parser register a translation error event?
+		if len(grpcroute.Spec.Hostnames) == 0 {
+			return nil
+		}
```

Comment:
Should this indeed skip the GRPCRoute? https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io/v1alpha2.GRPCRouteSpec indicates Hostname is optional, and doesn't mention special rules for empty-hostname GRPCRoutes.

AFAICT a GRPCRoute that specifies _nothing_ should then match any otherwise unmatched request. We'd want a test case for those also.

If we do fail here yeah, we should modify the signature so that the caller (a parser method) can emit a translation failure event.

Reply:
Do you think a rule with no match in a `GRPCRoute` with no hostnames should be a "catch-all" rule that matches ANY request? This behavior is not the same as the current implementation of `HTTPRoute`. I think the proper behavior is to keep the behavior the same as `HTTPRoute`: skip this `GRPCRoute` and report a translation error. If we want to change this behavior, we submit another PR.

Reply:
Yeah, this appears to be a regression introduced with combined services/routes for some reason. The spec definitely allows total wildcard hostnames in Listeners and they're an optional field on Route types that support them, so Routes that have no hostnames should match any-hostname requests that otherwise match their rules, inclusive of the total wildcard rule. That is the normal behavior for Kong routes with no hostname.

[generateKongRoutesFromHTTPRouteRule](https://github.com/Kong/kubernetes-ingress-controller/blob/main/internal/dataplane/parser/translate_httproute.go#L211-L285) via [ingressRulesFromHTTPRouteLegacyFallback](https://github.com/Kong/kubernetes-ingress-controller/blob/6f8ae7b27f26e3b5554366d83c2a24b1a6bc1575/internal/dataplane/parser/translate_httproute.go#L159-L187) would generate any-hostname rules, as it's just collecting the list of hostnames (which may be empty) before iterating over the rules.